dbms/src/Flash/Mpp/MPPTunnel.h

```diff
@@ -44,45 +44,164 @@ namespace DB
 {
 namespace tests
 {
-class MPPTunnelTest;
-class TestMPPTunnelBase;
+class TestMPPTunnel;
 } // namespace tests
 
 class EstablishCallData;
 
+enum class TunnelSenderMode
+{
+    SYNC_GRPC, // Using sync grpc writer
+    LOCAL, // Expose internal memory access, no grpc writer needed
+    ASYNC_GRPC // Using async grpc writer
+};
+
+/// TunnelSender is responsible for consuming data from Tunnel's internal send_queue and do the actual sending work
+/// After TunnelSend finished its work, either normally or abnormally, set ConsumerState to inform Tunnel
+class TunnelSender : private boost::noncopyable
+{
+public:
+    using MPPDataPacketPtr = std::shared_ptr<mpp::MPPDataPacket>;
+    using DataPacketMPMCQueuePtr = std::shared_ptr<MPMCQueue<MPPDataPacketPtr>>;
+    virtual ~TunnelSender() = default;
+    TunnelSender(TunnelSenderMode mode_, DataPacketMPMCQueuePtr send_queue_, PacketWriter * writer_, const LoggerPtr log_, const String & tunnel_id_)
+        : mode(mode_)
+        , send_queue(send_queue_)
+        , writer(writer_)
+        , log(log_)
+        , tunnel_id(tunnel_id_)
+    {
+    }
+    DataPacketMPMCQueuePtr getSendQueue()
+    {
+        return send_queue;
+    }
+    void consumerFinish(const String & err_msg);
+    String getConsumerFinishMsg()
+    {
+        return consumer_state.getMsg();
+    }
+    bool isConsumerFinished()
+    {
+        return consumer_state.msgHasSet();
+    }
+    const LoggerPtr & getLogger() const { return log; }
+    String getTunnelId()
+    {
+        return tunnel_id;
+    }
+
+protected:
+    /// TunnelSender use consumer state to inform tunnel that whether sender has finished its work
+    class ConsumerState
+    {
+    public:
+        ConsumerState()
+            : future(promise.get_future())
+        {
+        }
+
+        // before finished, must be called without protection of mu
```

Comment:
Looks like this comment is out of date, there is no `mu` in `TunnelSender`

Reply:
Yeah, I'll remove it.